File: sdks/go/qnsi/vault/vault.go
// Package vault is the QNSP Vault client — PQC-encrypted secret storage
// with versioning, rotation, and deletion.
package vault

import (
	"context"
	"net/http"
	"strconv"
	"time"

	"github.com/heossihq/qnsi-public/sdks/go/qnsi/internal/qnsicore"
)

const pathPrefix = "/proxy/vault/v1"

// Client is the QNSP vault client.
type Client struct {
	base *qnsicore.ServiceClient
}

// New constructs a vault client that shares an Activator and HTTP client
// with the parent qnsp.Client.
func New(act *qnsicore.Activator, httpClient *http.Client, timeout time.Duration) *Client {
	return &Client{base: qnsicore.NewServiceClient(act, httpClient, pathPrefix, timeout)}
}

// CreateSecretRequest is the JSON body for POST /vault/v1/secrets.
type CreateSecretRequest struct {
	Name       string         `json:"name"`
	PayloadB64 string         `json:"payloadB64"`
	Algorithm  string         `json:"algorithm,omitempty"`
	Metadata   map[string]any `json:"metadata,omitempty"`
}

func (c *Client) CreateSecret(ctx context.Context, req CreateSecretRequest, idempotencyKey string) (map[string]any, error) {
	// Wire contract: createSecretSchema = { tenantId(auto-injected), name(min3), payload, metadata,
	// rotationPolicy(default) }. Map the SDK's PayloadB64 -> payload; fold algorithm into metadata.
	// (Sending the raw struct put the value under "payloadB64", which the schema strips -> the
	// secret stored an empty payload.)
	metadata := map[string]any{}
	for k, v := range req.Metadata {
		metadata[k] = v
	}
	if req.Algorithm != "" {
		metadata["algorithm"] = req.Algorithm
	}
	body := map[string]any{"name": req.Name, "payload": req.PayloadB64}
	if len(metadata) > 0 {
		body["metadata"] = metadata
	}
	return c.base.Do(ctx, http.MethodPost, "/secrets", body, nil, idempotencyKey)
}

func (c *Client) GetSecret(ctx context.Context, secretID string) (map[string]any, error) {
	return c.base.Do(ctx, http.MethodGet, "/secrets/"+secretID, nil, nil, "")
}

func (c *Client) GetSecretVersion(ctx context.Context, secretID string, version int) (map[string]any, error) {
	return c.base.Do(ctx, http.MethodGet, "/secrets/"+secretID+"/versions/"+strconv.Itoa(version), nil, nil, "")
}

func (c *Client) RotateSecret(ctx context.Context, secretID string, payloadB64, algorithm string, idempotencyKey string) (map[string]any, error) {
	// Wire contract: rotateSecretSchema = { tenantId(auto), newPayload?, metadata?, rotationPolicy? }.
	// Map payloadB64 -> newPayload; fold algorithm into metadata.
	body := map[string]any{"newPayload": payloadB64}
	if algorithm != "" {
		body["metadata"] = map[string]any{"algorithm": algorithm}
	}
	return c.base.Do(ctx, http.MethodPost, "/secrets/"+secretID+"/rotate", body, nil, idempotencyKey)
}

func (c *Client) DeleteSecret(ctx context.Context, secretID string) error {
	_, err := c.base.Do(ctx, http.MethodDelete, "/secrets/"+secretID, nil, nil, "")
	return err
}

func (c *Client) ListSecretVersions(ctx context.Context, secretID string) (map[string]any, error) {
	return c.base.Do(ctx, http.MethodGet, "/secrets/"+secretID+"/versions", nil, nil, "")
}

File: sdks/rust/qnsi/src/vault.rs
//! QNSI Vault - PQC-encrypted secret storage with versioning, rotation,
//! and deletion.

use reqwest::Method;
use serde_json::Value;
use std::sync::Arc;

use crate::activation::Activation;
use crate::errors::Error;
use crate::service_client::ServiceClient;

const PATH_PREFIX: &str = "/proxy/vault/v1";

#[derive(Clone)]
pub struct Client {
    sc: ServiceClient,
}

impl Client {
    pub fn new(activation: Arc<Activation>, http: reqwest::Client) -> Self {
        Self { sc: ServiceClient::new(activation, http, PATH_PREFIX) }
    }

    pub async fn create_secret(
        &self,
        req: CreateSecretRequest,
        idempotency_key: Option<&str>,
    ) -> Result<Value, Error> {
        // Wire contract: createSecretSchema = { tenantId, name(min3), payload,
        // metadata, rotationPolicy(default) }. tenantId is injected by ServiceClient.
        // Map payload_b64 -> payload and fold algorithm into metadata (mirrors npm).
        let mut metadata = req.metadata.unwrap_or_default();
        if let Some(alg) = &req.algorithm {
            metadata.insert("algorithm".to_string(), Value::String(alg.clone()));
        }
        let body = serde_json::json!({
            "name": req.name,
            "payload": req.payload_b64,
            "metadata": metadata,
        });
        self.sc
            .request(Method::POST, "/secrets", Some(&body), None, idempotency_key)
            .await
    }

    pub async fn get_secret(&self, secret_id: &str) -> Result<Value, Error> {
        self.sc
            .request::<()>(Method::GET, &format!("/secrets/{}", secret_id), None, None, None)
            .await
    }

    pub async fn get_secret_version(&self, secret_id: &str, version: u64) -> Result<Value, Error> {
        self.sc
            .request::<()>(
                Method::GET,
                &format!("/secrets/{}/versions/{}", secret_id, version),
                None,
                None,
                None,
            )
            .await
    }

    pub async fn rotate_secret(
        &self,
        secret_id: &str,
        payload_b64: String,
        algorithm: Option<String>,
        idempotency_key: Option<&str>,
    ) -> Result<Value, Error> {
        // Wire contract: rotateSecretSchema = { tenantId, newPayload?, metadata?,
        // rotationPolicy? }. Map payload_b64 -> newPayload; algorithm into metadata.
        let mut body = serde_json::json!({ "newPayload": payload_b64 });
        if let Some(alg) = algorithm {
            body["metadata"] = serde_json::json!({ "algorithm": alg });
        }
        self.sc
            .request(
                Method::POST,
                &format!("/secrets/{}/rotate", secret_id),
                Some(&body),
                None,
                idempotency_key,
            )
            .await
    }

    pub async fn delete_secret(&self, secret_id: &str) -> Result<(), Error> {
        // tenantId is injected into the query centrally by ServiceClient.
        self.sc
            .request::<()>(Method::DELETE, &format!("/secrets/{}", secret_id), None, None, None)
            .await?;
        Ok(())
    }

    pub async fn list_secret_versions(&self, secret_id: &str) -> Result<Value, Error> {
        self.sc
            .request::<()>(
                Method::GET,
                &format!("/secrets/{}/versions", secret_id),
                None,
                None,
                None,
            )
            .await
    }
}

/// SDK-facing input for [`Client::create_secret`]. `create_secret` maps this to
/// the backend wire shape (`payload_b64` -> `payload`, `algorithm` folded into
/// `metadata`), so the field names here are the ergonomic SDK surface, not the wire.
#[derive(Debug, Clone, Default)]
pub struct CreateSecretRequest {
    pub name: String,
    pub payload_b64: String,
    pub algorithm: Option<String>,
    pub metadata: Option<serde_json::Map<String, Value>>,
}
